### DISCUSSION TITLE
Development experience: Native vs Web

### DISCUSSION THREAD

**Discussion by weiminghuaa** [MobileNativeFoundation/discussions#16] [General]
Without compare things like performance, etc. Why iOS community haven't provide some good web tools to implement hot reload, state manager? It's very helpful when I develop web or flutter app.

**Comment by zachgrayio** [comment]
Shots fired 😄

**Reply by weiminghuaa** [reply]
Let me declare that I am not trying to cause a war between native and web, but I really like these development tools and programming paradigm on the web.

**Reply by zachgrayio** [reply]
I totally agree with you for what it's worth, and am a huge fan of what flutter has been doing to move the needle in a dev experience in the mobile space on par with what modern web dev can look like; and its worked because we have seen both native android and iOS ui paradigms and experienced inspired by this come about. 

Would love to see some projects making additional impact on this front happen here

**Comment by keith** [comment]
I haven't personally tried this out but there is this project https://github.com/johnno1962/InjectionIII which I believe mains to cover what you're suggesting. Also things like this https://github.com/krzysztofzablocki/Playgrounds.

My perspective has been that projects like this can only cover N% of the problem, and unless N == 100 you row to not trust them and it's always simpler to not uses them. But I'd be interested to hear folks experiences who have used this more than I have.

**Reply by rudro** [reply]
I've tried Injection for Objective-C code and it works moderately well. As you note (with the N% comment) you need to be very aware of it's limitations and always watch out for the case where the updates don't work. It's not worked out so well for Swift code for me.

This topic is obviously a pretty big can of worms, but I think a quick answer might be that "Apple is trying but is clearly several years behind the web-dev experience". Apple would most likely point at dynamically updating SwiftUI previews and say that's one attempt at hot-reloading. 

In general I think it's a very hard problem to solve with a language like Swift, so the suggest would be to drop down to the `JavascriptCore` engine to get this done. At that point you're faced with either a significant loss in type-safety (most likely a lot of data shows up a dictionaries of basic types) or you need systems to constantly generate boilerplate to transform the data into real structs as it crosses the bridge (and even then you need to be careful that the changes to these don't constantly fire of recompilation of the Swift code). There are a lot more details to discuss, but there are some quick ones that come to mind.
